package/MDAnalysis/analysis/base.py

```diff
@@ -430,25 +517,219 @@ def run(self, start=None, stop=None, step=None, frames=None,
         verbose = getattr(self, '_verbose',
                           False) if verbose is None else verbose
 
-        self._setup_frames(self._trajectory, start=start, stop=stop,
-                           step=step, frames=frames)
         logger.info("Starting preparation")
-        self._prepare()
         logger.info("Starting analysis loop over %d trajectory frames",
                     self.n_frames)
 
-        for i, ts in enumerate(ProgressBar(
-                self._sliced_trajectory,
+        if len(indexed_frames) == 0: # if `frames` were empty in `run` or `stop=0`
+            return self
+        
+        frame_indices, frames = indexed_frames[:, 0], indexed_frames[:, 1], 
+        trajectory = self._trajectory[frames]
+        for idx, ts in enumerate(ProgressBar(trajectory,
                 verbose=verbose,
                 **progressbar_kwargs)):
+            i = frame_indices[idx]
             self._frame_index = i
             self._ts = ts
             self.frames[i] = ts.frame
             self.times[i] = ts.time
             self._single_frame()
         logger.info("Finishing up")
+        return self
+
+    def _setup_computation_groups(self, n_parts: int, start=None, stop=None, step=None, frames=None) -> list[np.ndarray]: 
+        """
+        Set the self._computation_groups (the workload distribution scheme) for the future delayed
+        computations.
+
+        Parameters
+        ----------
+        start : int, optional
+            start frame of analysis
+        stop : int, optional
+            stop frame of analysis
+        step : int, optional
+            number of frames to skip between each analysed frame
+        frames : array_like, optional
+            array of integers or booleans to slice trajectory; `frames` can
+            only be used *instead* of `start`, `stop`, and `step`. Setting
+            *both* `frames` and at least one of `start`, `stop`, `step` to a
+            non-default value will raise a :exc:`ValueError`.
+
+        Raises
+        ------
+        ValueError
+            if *both* `frames` and at least one of `start`, `stop`, or `frames`
+            is provided (i.e., set to another value than ``None``)
+
+        Returns
+        -------
+        computation_groups : list of (n, 2) shaped np.ndarrays with frame indices and numbers
```

Comment:
IIRC, the `Returns` docstring has a different syntax. Pleas double check.
```suggestion
        list
            (n, 2)-shaped np.ndarray with frame indices and numbers
```

Reply:
Thanks, it was the autodoc automatic format that I didn't check. Fixed.